File: Python/Stack/M-739. Daily Temperatures.py
class Solution:
    @classmethod
    def dailyTemperatures(self, temperatures: list[int]) -> list[int]:
        stack = []
        result = [0] * len(temperatures)
        for i in range(len(temperatures) - 1, -1, -1):
            while len(stack) and temperatures[i] >= temperatures[stack[len(stack) - 1]]:
                stack.pop()
            if not len(stack):
                result[i] = 0
            else:
                result[i] = stack[len(stack) - 1] - i
            stack.append(i)
        return result


testData = [
    {
        "temperatures": [73, 74, 75, 71, 69, 72, 76, 73],
        "ans": [1, 1, 4, 2, 1, 1, 0, 0],
    },
    {
        "temperatures": [30, 40, 50, 60],
        "ans": [1, 1, 1, 0],
    },
    {
        "temperatures": [30, 60, 90],
        "ans": [1, 1, 0],
    },
]

for data in testData:
    print(data["ans"] == Solution.dailyTemperatures(data["temperatures"]))

File: Javascript/Stack/M-739. Daily Temperatures.ts
function dailyTemperatures(temperatures: number[]): number[] {
  const stack: number[] = [];
  const result: number[] = new Array(temperatures.length).fill(0);
  for (let i = temperatures.length - 1; i >= 0; i--) {
    while (
      stack.length &&
      temperatures[i] >= temperatures[stack[stack.length - 1]]
    ) {
      stack.pop();
    }
    if (!stack.length) result[i] = 0;
    else result[i] = stack[stack.length - 1] - i;
    stack.push(i);
  }
  return result;
}

const testData = [
  {
    temperatures: [73, 74, 75, 71, 69, 72, 76, 73],
    ans: [1, 1, 4, 2, 1, 1, 0, 0],
  },
  {
    temperatures: [30, 40, 50, 60],
    ans: [1, 1, 1, 0],
  },
  {
    temperatures: [30, 60, 90],
    ans: [1, 1, 0],
  },
];

for (const data of testData) {
  console.log(
    JSON.stringify(data['ans']) ===
      JSON.stringify(dailyTemperatures(data['temperatures']))
  );
}
